Insight streams tutorial (#61)
redis-stack/redis-stack-docs: docs/stack/insight/tutorials/insight-stream-consumer.md

---
title: "Manage streams and consumer groups in RedisInsight"
linkTitle: Streams
description: Learn how to manage streams and consumer groups in RedisInsight
weight: 1
---

A _stream_ is an append-only log file. 
When you add data to it, you cannot change it. 
That may seem like a disadvantage; however, a stream serves as a log or single source of truth. 

> It is used as a buffer between processes that work at different speeds and do not need to know about each other. 

Reviewer: ```suggestion
It can also be used as a buffer between processes that work at different speeds and do not need to know about each other. 
```

Suggesting rewording as this is a second, distinct, use case.

For more conceptual information about streams, see [Redis Streams](/docs/manual/data-types/streams). 

In this topic, you will learn how to add and work with streams as well as consumer groups in RedisInsight.

Here's a stream that models temperature and humidity sensors. 
In this stream, sensors are considered _producers_, which broadcast data. 
The point of a stream is that it's not going to end, so you cannot capture whole datasets and do some processing on them. 

> In order to process the data, a stream is separated into different roles. 

Reviewer: ```suggestion
Processes interacting with the stream perform one of two roles:```

The two roles are "consumer" and "producer".  I think there needs to be a sentence here about the producer role before we go into the consumer role?

A _consumer_ reads from the stream and does some work on it. 
For example, if the temperature is above a certain threshold, it puts a message out to turn on the air conditioner in that unit or notify the maintenance.

<img src="../images/stream.png">

It is possible to have multiple consumers doing different jobs, one measuring humidity, and another taking temperature measurements over periods of time. 

> Redis runs everything in memory that is finite. 

Reviewer: ```suggestion
Redis stores a copy of the entire dataset in memory, which is a finite resource.
```

To avoid runaway data, streams can be trimmed when you add something to them. 

> You can specify to remove everything but a number of last entries or remove stream that has an ID lower than specified, for example a timestamp. 

Reviewer: ```suggestion
When adding to a stream with `XADD`, you can optionally specify that the stream should be trimmed to a specific or approximate number of the newest entries, or to only include entries whose ID is higher than the ID specified.
```

Not sure I am 100% on my wording here, but I tried to clarify the options.

> You can also expire streams using keys by writing to a different key each day and expiring those keys after a period of time, say a week. 

Reviewer: ```suggestion
You can also manage the storage required for streaming data using key expiry.  For example, by writing each day's data to its own stream in Redis and expiring each stream's key after a period of time, say a week.
```

> An ID can be any number.

Reviewer: ```suggestion
An ID can be any number, but each new entry in the stream must have an ID whose value is higher than the last ID added to the stream.
```

## Adding new entries 

> Use `XADD` with `*` for ID and that automatically generates timestamps plus fields and values. There are a couple of ways of retrieving things. You could give it a time range or you could ask for everything that's happened since a timestamp or ID that you specify. On one command you can ask for anything from 10:30 until 11:15 am on each day over a specified period of time.

Reviewer: ```suggestion
Use `XADD` with `*` for the ID to have Redis automatically generate a new ID for you consisting of a millisecond precision timestamp, a dash and a sequence number.  For example `1656416957625-0`.  Then supply the field names and values to store in the new stream entry.

There are a couple of ways of retrieving things. You can retrieve entries by time range or you could ask for everything that's happened since a timestamp or ID that you specify. Using a single command you can ask for anything from 10:30 until 11:15 am on a given day.
```

## Consumer groups

A more realistic use case would be a system with many temperature sensors whose data Redis puts in a stream, records the time they arrive, and orders them. 

<img src="../images/consumer.png">

On the right side we have two consumers that read the stream. One of them is alerting if the temperature is over a certain number and texting the maintenance crew that they need to do something, and the other is a data warehouse that is taking the data and putting it into a database. 

They run independently of each other. 
Up in the right, we have another sort of task. 
Let's assume that alerting and data warehouse are really fast. 
You get a message whether the temperature is larger than a specific value, which might take a millisecond. 
And alerting can keep up with the data flow. 
One way you can scale consumers is _consumer groups_, which allows multiple instances of the same consumer or same code to work as a team to process the stream.

## Managing streams in RedisInsight

> You can add a stream in RedisInsight in two ways: create a new stream or run an existing stream.

Reviewer: ```suggestion
You can add a stream in RedisInsight in two ways: create a new stream or add to an existing stream.
```

To create a stream, start by selecting the key type (stream). 

> You cannot set time to live (TTL) because it cannot be put on a message in a stream; it can only be done on a Redis key. Name the stream _mystream_. 

Reviewer: You can set TTL on the new stream key, you can't set TTL on an entry in a stream though.

<img width="725" alt="Screen Shot 2022-06-28 at 12 52 17 PM" src="https://user-images.githubusercontent.com/94062/176171877-76fb4bcf-9b2a-400b-afab-706e847afb36.png">
.

Then, set the *Entry ID* to `*` to default to timestamp. 

> If this were a supermarket queue, for example, you would set it to a different number. 

Reviewer: Not sure that this fits here, could be omitted?  Or changed to "If you have your own ID generation strategy, enter the next ID from your sequence.  Remember that the ID must be higher than the ID of any other entry in the stream."

In this example, we want to track the entries by time. Then, enter fields and values using + to add more than one (for example, name: Simon; location: England). 
Now you have a stream that appears in the **Streams** view and you can continue adding fields and values to it.

> RedisInsight runs read commands for you so you can see them run in the **Streams** view. 

Reviewer: ```suggestion
RedisInsight runs read commands for you so you can see the stream entries in the **Streams** view. 
```

> And the **Consumer Groups** view shows each of consumers and the last time Redis allocated a message, what the idea of it was and how many times that process has happened, and whether you have told Redis that you are finished working with that task.

Reviewer: ```suggestion
And the **Consumer Groups** view shows each consumers in a given consumer group and the last time Redis allocated a message, what the ID of it was and how many times that process has happened, and whether a consumer has you have told Redis that you are finished working with that task using the `XACK` command.
```

> streams@1.0.0 consumergroup
> node consumer_group.js -- "consumer1"

Starting consumer consumer1...
Consumer group temphumidity_consumers exists, not created.
Reading stream...
Received entry 1632771059039-0:
[ 'location', '96', 'temp', '15.4', 'humidity', '70' ]
Acknowledged processing of entry 1632771059039-0.
Reading stream...
{{< / highlight >}}

In a second terminal:

{{< highlight bash >}}
npm run consumergroup consumer2
{{< / highlight >}}

And in a third:

{{< highlight bash >}}
npm run consumergroup consumer3
{{< / highlight >}}

The consumers will run indefinitely, waiting for new messages to be added to the stream by a producer instance when they have collectively consumed the entire stream. 
Note that in this model, each consumer instance does not receive all of the entries from the stream, but the three members of the group each receive a subset.

### View the stream in RedisInsight

1. Launch RedisInsight.
2. Select `localhost:6379`
3. Select **STREAM**. Optionally, select full screen from the upper right corner to expand the view.

<img src="../images/insight_streams.png">

You can now toggle between **Stream** and **Consumer Groups** views to see your data. 

> As mentioned earlier in this topic, a stream is an append-only log so you can delete messages. 

Reviewer: ```suggestion
As mentioned earlier in this topic, a stream is an append-only log so you can't mofify the contents of an entry, but you can delete an entire entry. 
```

A case when that's useful is in the event of a so-called _poison-pill message_ that can cause consumers to crash. You can physically remove such messages in the **Streams** view or use the `XDEL` command at the command-line interface (CLI).

> You can continue interacting with your stream at the CLI. For example, to get the number of entries per minute, enter `XLEN` command:

Reviewer: ```suggestion
You can continue interacting with your stream at the CLI. For example, to get the current length of a stream, use the `XLEN` command:
```
